### PR TITLE
Iox #14 refactor chunk settings into struct

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer

This is just a "small" refactoring from the last PR. I created a `ChunkSettings` class and used it for `MemoryManager::getChunk` and the `ChunkHeader` ctor.

The code from `MemoryManager::requiredChunkSize` also moved to the new class, together with the tests. This makes the PR so large.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Partially closes #14

### CODE REVIEW

File: iceoryx_binding_c/test/moduletests/test_event_info.cpp [hunk 1]
```diff
@@ -78,10 +78,16 @@ class iox_event_info_test : public Test
...
-                                        iox::CHUNK_DEFAULT_PAYLOAD_ALIGNMENT,
-                                        iox::CHUNK_NO_CUSTOM_HEADER_SIZE,
-                                        iox::CHUNK_NO_CUSTOM_HEADER_ALIGNMENT);
+
+        auto chunkSettingsResult = ChunkSettings::create(PAYLOAD_SIZE, iox::CHUNK_DEFAULT_PAYLOAD_ALIGNMENT);
+        EXPECT_FALSE(chunkSettingsResult.has_error());
+        if (chunkSettingsResult.has_error())
+        {
+            return nullptr;
+        }
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/test_event_info.cpp:87]: unfortunately I couldn't use `ASSERT_THAT` since it would return with `void`. Preciously, there was just a `terminate` in the code when wrong parameter were specified.

File: iceoryx_posh/source/mepoo/chunk_settings.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        return cxx::error<ChunkSettings::Error>(ChunkSettings::Error::REQUIRED_CHUNK_SIZE_EXCEEDS_MAX_CHUNK_SIZE);
+    }
+
+    return cxx::success<ChunkSettings>(ChunkSettings{payloadSize,
+                                                     adjustedPayloadAlignment,
+                                                     customHeaderSize,
+                                                     adjustedCustomHeaderAlignment,
+                                                     static_cast<uint32_t>(requiredChunkSize)});
+}
+uint64_t ChunkSettings::calculateRequiredChunkSize(const uint32_t payloadSize,
```
Comment (elBoberido) [inline iceoryx_posh/source/mepoo/chunk_settings.cpp:79]: this is basically `MemoryManager::requiredChunkSize`

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp [hunk 3]
```diff
@@ -316,180 +316,44 @@ INSTANTIATE_TEST_CASE_P(ChunkHeader_test,
...
-TEST_P(ChunkHeader_AlteringPayloadWithCustomHeader,
-       checkIntegrityOfChunkHeader_CustomHeader_SizeGreaterThanChunkHeader_AlignmentEqualsToChunkHeaderAlignment)
-{
-    const auto payloadParams = GetParam();
-
-    constexpr uint32_t CUSTOM_HEADER_SIZE{sizeof(ChunkHeader) * 2U};
-    constexpr uint32_t CUSTOM_HEADER_ALIGNMENT{alignof(ChunkHeader)};
+    for (const auto customHeaderAlignment : CUSTOM_HEADER_ALIGNMENTS)
+    {
+        SCOPED_TRACE(std::string("Custom header alignment = ") + std::to_string(customHeaderAlignment));
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp:332]: this was done to reduce code duplication

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    const auto payload = GetParam();
+
+    SCOPED_TRACE(std::string("Payload size = ") + std::to_string(payload.size) + std::string("; alignment = ")
+                 + std::to_string(payload.alignment));
+
+    constexpr uint32_t CUSTOM_HEADER_SIZES[]{
+        1U, sizeof(ChunkHeader) / 2U, sizeof(ChunkHeader), sizeof(ChunkHeader) * 2U};
+    constexpr uint32_t CUSTOM_HEADER_ALIGNMENTS[]{0U, 1U, alignof(ChunkHeader), alignof(ChunkHeader)};
+
+    for (const auto customHeaderAlignment : CUSTOM_HEADER_ALIGNMENTS)
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:322]: this too

File: iceoryx_posh/test/moduletests/test_mepoo_memory_manager.cpp [hunk 5]
```diff
@@ -364,434 +396,4 @@ TEST_F(MemoryManager_test, addMemPoolWithChunkCountZeroShouldFail)
     EXPECT_DEATH({ sut->configureMemoryManager(mempoolconf, allocator, allocator); }, ".*");
 }
 
-TEST_F(MemoryManager_test, requiredChunkSize_AllParameterMinimal_ResultsIn_SizeOfChunkHeader)
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_mepoo_memory_manager.cpp:367]: these are moved to the `ChunkSettings` tests

File: iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp [hunk 6]
```diff
@@ -268,6 +268,15 @@ constexpr bool isCompiledOn32BitSystem()
     return INTPTR_MAX == INT32_MAX;
 }
 
+/// @brief Checks if an unsigned integer is a power of two
+/// @return true if power of two, otherwise false
+template <typename T>
+constexpr bool isPowerOfTwo(const T n)
+{
+    static_assert(std::is_unsigned<T>::value && !std::is_same<T, bool>::value, "Only unsigned integer are allowed!");
+    return n && ((n & (n - 1U)) == 0U);
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp:277]: for the mathematicians to prove ;)

**Review by elBoberido — COMMENTED** [review]
I'm a bit hesitant to use the `ChunkSettings` also for `ChunkSender` and `PublisherPortUser`. What do you think?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 7]
```diff
@@ -195,7 +195,7 @@ namespace roudi
 {
 using ConfigFilePathString_t = cxx::string<1024>;
 
-constexpr char IPC_CHANNEL_ROUDI_NAME[] = "/roudi";
+constexpr char IPC_CHANNEL_ROUDI_NAME[] = "roudi";
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:198]: noticed this while I was updating iceray yesterday

File: iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl [hunk 8]
```diff
@@ -51,8 +51,11 @@ inline cxx::expected<ChunkManagement*, TypedMemPoolError> TypedMemPool<T>::acqui
         return cxx::error<TypedMemPoolError>(TypedMemPoolError::FatalErrorReachedInconsistentState);
     }
 
-    new (chunkHeader) ChunkHeader(
-        m_memPool.getChunkSize(), sizeof(T), alignof(T), CHUNK_NO_CUSTOM_HEADER_SIZE, CHUNK_NO_CUSTOM_HEADER_ALIGNMENT);
+    auto chunkSettingsResult = mepoo::ChunkSettings::create(sizeof(T), alignof(T));
+    // this is safe since we use correct values for size and alignment
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl:55]: Then please add `cxx::Expects` for the unlikely case that we change the requirements for correct size and alignment later and this goes south.

File: iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl [hunk 9]
```diff
@@ -123,10 +126,12 @@ inline uint32_t TypedMemPool<T>::getUsedChunks() const noexcept
 template <typename T>
 inline uint64_t TypedMemPool<T>::requiredChunkSize() noexcept
 {
+    auto chunkSettingsResult = mepoo::ChunkSettings::create(sizeof(T), alignof(T));
+    // this is safe since we use correct values for size and alignment
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl:130]: Please add `cxx::Expects`.

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+
+            auto sutResult =
+                ChunkSettings::create(payload.size, payload.alignment, customHeaderSize, customHeaderAlignment);
+            ASSERT_FALSE(sutResult.has_error());
+            auto& sut = sutResult.value();
+
+            EXPECT_THAT(sut.requiredChunkSize(), Eq(expectedSize));
+        }
+    }
+}
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:346]: We need one more tests which calls `ChunkSettings::customHeaderAlignment`. Maybe you could add this to an already existing test.

File: iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp (hunk 6 above)
Comment (elfenpiff) [reply iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp:277]: 1. n != 0 so the first part is true. 
   * if n is a power of two only one 1 (at position k) can be found in the bit representation, everything else is 0
   * in n - 1 there are all bits set to 1 which are at a position less then k.
   * example: `1 0 0 0 & 0 1 1 1 = 0 0 0 0 == 0U`
2. n == 0 first part is false ... all fine

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl (hunk 8 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/mepoo/typed_mem_pool.inl:55]: I had it first, but decided to remove it since `sizeof` and `alignof` is used and I wouldn't be able to test the `cxx::Expects`

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+       MultipleCustomHeaderSizesAndAlignments_ResultsIn_RequiredChunkSizeIsCorrect)
+{
+    const auto payload = GetParam();
+
+    SCOPED_TRACE(std::string("Payload size = ") + std::to_string(payload.size) + std::string("; alignment = ")
+                 + std::to_string(payload.alignment));
+
+    constexpr uint32_t CUSTOM_HEADER_SIZES[]{
+        1U, sizeof(ChunkHeader) / 2U, sizeof(ChunkHeader), sizeof(ChunkHeader) * 2U};
+    constexpr uint32_t CUSTOM_HEADER_ALIGNMENTS[]{0U, 1U, alignof(ChunkHeader), alignof(ChunkHeader)};
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:320]: Why did you add ``alignof(ChunkHeader)`` twice?

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp (hunk 10 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:393]: done

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp (hunk 11 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:367]: because I'm obviously dumb and meant `alignof(ChunkHeader) / 2U`

Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp]: fixed

**Review by elfenpiff — DISMISSED** [review]

File: iceoryx_binding_c/test/moduletests/test_subscriber.cpp [hunk 12]
```diff
@@ -82,10 +82,16 @@ class iox_sub_test : public Test
     iox::mepoo::SharedChunk getChunkFromMemoryManager()
     {
         constexpr uint32_t PAYLOAD_SIZE{100U};
-        return m_memoryManager.getChunk(PAYLOAD_SIZE,
-                                        iox::CHUNK_DEFAULT_PAYLOAD_ALIGNMENT,
-                                        iox::CHUNK_NO_CUSTOM_HEADER_SIZE,
-                                        iox::CHUNK_NO_CUSTOM_HEADER_ALIGNMENT);
+
+        auto chunkSettingsResult = ChunkSettings::create(PAYLOAD_SIZE, iox::CHUNK_DEFAULT_PAYLOAD_ALIGNMENT);
+        EXPECT_FALSE(chunkSettingsResult.has_error());
```
Comment (mossmaurice) [inline iceoryx_binding_c/test/moduletests/test_subscriber.cpp:87]: `ASSERT_FALSE`?

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp [hunk 13]
```diff
@@ -316,180 +316,44 @@ INSTANTIATE_TEST_CASE_P(ChunkHeader_test,
                             PayloadParams{sizeof(PayloadOffset_t), alignof(PayloadOffset_t) * 2},
                             PayloadParams{sizeof(PayloadOffset_t) * 42U, alignof(PayloadOffset_t) * 2}));
 
-// BEGIN ALTERING CUSTOM HEADER SIZE WITH ALIGNMENT EQUAL TO ONE
-
-TEST_P(ChunkHeader_AlteringPayloadWithCustomHeader,
-       checkIntegrityOfChunkHeader_CustomHeader_SizeEqualsToOne_AlignmentEqualsToOne)
+TEST_P(ChunkHeader_AlteringPayloadWithCustomHeader, checkIntegrityOfChunkHeaderWithCustomHeader)
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_header.cpp:319]: Nitpick: Test case names shall start with capital letter

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+{
+using namespace ::testing;
+
+using iox::mepoo::ChunkHeader;
+using iox::mepoo::ChunkSettings;
+using PayloadOffset_t = iox::mepoo::ChunkHeader::PayloadOffset_t;
+
+// BEGIN GETTER METHOD TESTS
+
+TEST(ChunkSettings_test, payloadSizeReturnsCorrectValue)
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:32]: Nitpick: same here for all test cases, capital letter

File: iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+#include <cstdint>
+
+namespace iox
+{
+namespace mepoo
+{
+class ChunkSettings
+{
+  public:
+    enum class Error
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp:32]: Nitpick: Other classes using `expected` define this outside of the class and call it `ChunkSettingsError`

**Review by mossmaurice — COMMENTED** [review]
just one file left

File: iceoryx_binding_c/test/moduletests/test_subscriber.cpp (hunk 12 above)
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_subscriber.cpp:87]: not possible, see https://github.com/eclipse-iceoryx/iceoryx/pull/629#discussion_r599545927

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp (hunk 14 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:32]: in this case it's the name of the method, so I thought it's better this way. Do you still want me to change it?

File: iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp (hunk 15 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp:32]: yes, but I wanted to test the approach with the nested class and I find it better since its more clear and one change the name of the class without forgetting to change the name of the enum. I can change it if it bothers you much.

File: iceoryx_binding_c/test/moduletests/test_event_info.cpp (hunk 1 above)
Comment (MatthiasKillat) [reply iceoryx_binding_c/test/moduletests/test_event_info.cpp:87]: Preciously ... Gollum...

If I understand correctly it is because you have to return something here even in the error case. That is ok, since the test will fail due to the ``EXPECT_THAT``

File: iceoryx_dds/test/mocks/chunk_mock_dds.hpp [hunk 16]
```diff
@@ -49,8 +50,11 @@ class ChunkMockDDS
 
         memset(m_rawMemory, 0xFF, Size);
 
-        m_chunkHeader = new (m_rawMemory) iox::mepoo::ChunkHeader(
-            Size, sizeof(T), alignof(T), iox::CHUNK_NO_CUSTOM_HEADER_SIZE, iox::CHUNK_NO_CUSTOM_HEADER_ALIGNMENT);
+        auto chunkSettingsResult = iox::mepoo::ChunkSettings::create(sizeof(T), alignof(T));
+        iox::cxx::Ensures(!chunkSettingsResult.has_error() && "Invalid parameter for ChunkMock");
+        auto& chunkSettings = chunkSettingsResult.value();
+
+        m_chunkHeader = new (m_rawMemory) iox::mepoo::ChunkHeader(Size, chunkSettings);
```
Comment (MatthiasKillat) [inline iceoryx_dds/test/mocks/chunk_mock_dds.hpp:57]: Is m_rawMemory large enough to construct the header in (also considering alignment)?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp [hunk 17]
```diff
@@ -37,6 +37,7 @@ enum class AllocationError
     RUNNING_OUT_OF_CHUNKS,
     TOO_MANY_CHUNKS_ALLOCATED_IN_PARALLEL,
     INVALID_CHUNK,
+    INVALID_PARAMETER_FOR_CHUNK,
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp:40]: PARAMETER or better INVALID_CHUNK_SETTINGS? What is the error used for?

File: iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp (hunk 15 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/mepoo/chunk_settings.hpp:32]: public nested class is preferable since it ties the error to the ChunkSettings..

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp (hunk 14 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:32]: We are not consistent in this way as of now and I do not know of any agreement. I favor lower case names for the same reason.
BTW: I like lower case for private API and upper case for public API. Subtle, easy on the eye, but noticeable (mainly interesting if public methods are called to implement private ones).

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+    auto& sut = sutResult.value();
+
+    EXPECT_THAT(sut.requiredChunkSize(), Eq(EXPECTED_SIZE));
+}
+
+// END GETTER METHOD TESTS
+
+// BEGIN EXCEEDING CHUNK SIZE TESTS
+
+TEST(ChunkSettings_test, NoCustomPayloadAlignmentAndTooLargePayload_Fails)
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:113]: Here the test names are lower case. I do not care much either way but it should be consistent within the file.

File: iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp (hunk 6 above)
Comment (MatthiasKillat) [reply iceoryx_utils/include/iceoryx_utils/cxx/helplets.hpp:277]: We already had this @elBoberido and it works.
The above is no full proof, but here you go.
Assume n != 0, it is handled in the first operand of &&.
If is a power of two it is trivial since exactly one bit is 1 and after subtraction it is 0 and the lower bits are 1 and so the result of & is 0.
Otherwise at least two bits are 1 and subtraction of 1 will leave the highest bit unchanged, causing the result of & to be !=0. (subtraction will change all the bits lower than the lowest bit set to 1 and the lowest set bit to 0 but no more)

A related task (more general):
Now count the number of set bits k in a n bit number but the algorithm only is allowed to take time O(k), i.e. O(n) is not sufficient (and does not need much insight, hence is boring).

File: iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp (hunk 14 above)
Comment (mossmaurice) [reply iceoryx_posh/test/moduletests/test_mepoo_chunk_settings.cpp:32]: I'd tend the other way. I'd rather rename the test case to something like `CallingPayloadSizeMethodReturnsCorrectValue` and keep all unit test cases upper case. But that's just what makes it easier for me to read, no hard feelings here.

File: iceoryx_binding_c/test/moduletests/test_event_info.cpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_event_info.cpp:87]: Exactly. In this case an empty `SharedChunk` will be created and depending on the test this might lead to a termination. I could also just terminate in this case but it should never happen since the `ChunkSettings` parameter are valid.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp (hunk 17 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp:40]: it's used for cases where invalid parameter are provided for the chunk e.g. payload alignment

File: iceoryx_dds/test/mocks/chunk_mock_dds.hpp (hunk 16 above)
Comment (elBoberido) [reply iceoryx_dds/test/mocks/chunk_mock_dds.hpp:57]: yes, it's large enough and also aligned. See lines 96 and 97. This mock doesn't use the custom header, therefore the payload is always adjacent to the `ChunkHeader`.

But I messed up the `ChunkHeader` tests. I need to provide some memory where the `ChunkHeader` is constructed, since it will write outside of it's bounds if the back-offset is necessary, e.g. when the payload alignment is larger than the `ChunkHeader` alignment. Will fix that.

Comment (elBoberido) [reply iceoryx_dds/test/mocks/chunk_mock_dds.hpp:57]: Forget it. I didn't mess up and did the right thing. Just forgot that it's not in the fixture but in `createChunksOnMultipleAddresses`

**Review by elfenpiff — DISMISSED** [review]

**Review by elfenpiff — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
